datasets/name_list.py

```diff
@@ -56,24 +60,47 @@ def generate_examples(self, num_examples):
                 evaluation_fn=self.evaluate_correct,
                 number_of_questions=self.count_questions(is_question),
                 is_question=is_question,
+                reset_message=self.reset_message
             )
             examples.append(example)
         return examples
 
     def evaluate_correct(
         self, questions: List[str], responses: List[str], expected_answers: List[str]
     ) -> Tuple[int, int, List[str]]:
-        not_found = []
-        for name in expected_answers:
-            if not name.lower() in responses[-1].lower():
-                not_found.append(name)
-        count = len(expected_answers) - len(not_found)
-        if len(not_found) == 0:
-            reasoning = "All expected names were found in the response."
-        else:
-            not_found = ", ".join(not_found)
-            reasoning = f"{count} names out of {len(expected_answers)} were found. Names {not_found} were not in the response."
-        return count, len(expected_answers), [reasoning]
+        try:
+            reasoning = []
+            answer_items = [n.lower() for n in sanitize_and_parse_json(responses[0])]
+            correct = 0
+            penalties = 0
+            not_found = []
+            lowered_expected = [n.lower() for n in expected_answers]
+
+            # Check answer -> expected
+            for name in answer_items:
+                if name in lowered_expected:
+                    correct += 1
+                else:
+                    penalties += 1
+                    reasoning.append(f"Name: {name} not expected.")
```

Comment:
This does not consider duplicates (e.g. it could score double by saying each name twice) and it seems to allow negative scores too. I'd suggest to remove a name from the list once is found and to set a minimum score of zero.

Reply:
Good point on the duplicates. Negative scores are somewhat intended? I don't know, I think there is a difference between just not remembering the correct information and actively performing badly through hallucinations.